packages/contracts-bedrock/src/Safe/LivenessModule.sol

```diff
@@ -0,0 +1,183 @@
+        _verifyGuard();
+
+        // Check that the owner to remove has not signed a transaction in the last 30 days
+        require(
+            livenessGuard.lastLive(owner) < block.timestamp - livenessInterval,
+            "LivenessModule: owner has signed recently"
+        );
+
+        // Calculate the new threshold
+        address[] memory owners = safe.getOwners();
+        uint256 numOwners = owners.length - 1;
+        uint256 thresholdAfter;
+        if (hasMinOwners(numOwners)) {
+            // Call the Safe to remove the owner and update the threshold
+            thresholdAfter = get75PercentThreshold(numOwners);
+            address prevOwner = _getPrevOwner(owner, owners);
+            _removeOwner({ _prevOwner: prevOwner, _owner: owner, _threshold: thresholdAfter });
+        } else {
+            // The number of owners is dangerously low, so we wish to transfer the ownership of this Safe
+            // to the fallback owner.
+
+            // Remove owners one at a time starting from the last owner.
+            // Since we're removing them in order from last to first, the ordering will remain constant,
+            // and we shouldn't need to query the list of owners again.
+            for (uint256 i = owners.length - 1; i >= 0; i--) {
+                address currentOwner = owners[i];
+                address prevOwner = _getPrevOwner(currentOwner, owners);
+                if (currentOwner != address(this)) {
+                    // Call the Safe to remove the owner
+                    _removeOwner({ _prevOwner: prevOwner, _owner: currentOwner, _threshold: 1 });
+                }
+            }
+
+            // Add the fallback owner as the sole owner of the Safe
+            _giveToFallbackOwner();
+        }
+
+        _verifyFinalState();
+    }
+
+    /// @notice Sets the fallback owner as the sole owner of the Safe with a threshold of 1
+    function _giveToFallbackOwner() internal {
+        safe.execTransactionFromModule({
+            to: address(safe),
+            value: 0,
+            operation: Enum.Operation.Call,
+            data: abi.encodeCall(OwnerManager.addOwnerWithThreshold, (fallbackOwner, 1))
+        });
+    }
+
+    /// @notice Removes the owner `owner` from the Safe and updates the threshold to `_threshold`.
+    /// @param _prevOwner Owner that pointed to the owner to be removed in the linked list
+    /// @param _owner Owner address to be removed.
+    /// @param _threshold New threshold.
+    function _removeOwner(address _prevOwner, address _owner, uint256 _threshold) internal {
+        safe.execTransactionFromModule({
+            to: address(safe),
+            value: 0,
+            operation: Enum.Operation.Call,
+            data: abi.encodeCall(OwnerManager.removeOwner, (_prevOwner, _owner, _threshold))
+        });
+    }
+
+    /// @notice A FREI-PI invariant check enforcing requirements on number of owners and threshold.
+    function _verifyFinalState() internal view {
```

Comment:
Having fuzz tests over `removeOwner` with varying sizes of safes would make me feel much better about this function

Reply:
Would that test be satisfied by what I've spec'd here? https://github.com/ethereum-optimism/optimism/pull/7540#discussion_r1370276542

Reply:
Resolving, this is tracked in the PR description.

Reply:
This is tracked in the PR description checklist. Resolving.